File: exercise/week04/c#/Construct/ArtefactUpdaterFactory.cs
namespace Construct;

public static class ArtefactUpdaterFactory
{
    private const string AgedSignal = "Aged Signal";
    private const string BackdoorPass = "Backdoor Pass to TAFKAL80ETC Protocol";
    private const string SulfurasCore = "Sulfuras Core Fragment";
    private const string StealthCloak = "Stealth Cloak v2.0";

    //static method to return the correct updater based on the artefact name
    public static IArtefactUpdater For(string artefactName)
        => artefactName switch
        {
            AgedSignal => new AgedSignalUpdater(),
            BackdoorPass => new BackdoorPassUpdater(),
            SulfurasCore => new SulfurasUpdater(),
            StealthCloak => new StealthCloakUpdater(),
            _ => new NormalArtefactUpdater()
        };
}

File: exercise/week04/ts/src/construct/artefactUpdaterFactory.ts
import { ArtefactUpdater } from "./artefactUpdater";
import { NormalArtefactUpdater } from "./normalArtefactUpdater";
import { AgedSignalUpdater } from "./agedSignalUpdater";
import { BackdoorPassUpdater } from "./backdoorPassUpdater";
import { SulfurasUpdater } from "./sulfurasUpdater";
import { StealthCloakUpdater } from "./stealthCloakUpdater";

const AGED_SIGNAL = "Aged Signal";
const BACKDOOR_PASS = "Backdoor Pass to TAFKAL80ETC Protocol";
const SULFURAS_CORE = "Sulfuras Core Fragment";
const STEALTH_CLOAK = "Stealth Cloak v2.0";

export function updaterFor(name: string): ArtefactUpdater {
    switch (name) {
        case AGED_SIGNAL: return new AgedSignalUpdater();
        case BACKDOOR_PASS: return new BackdoorPassUpdater();
        case SULFURAS_CORE: return new SulfurasUpdater();
        case STEALTH_CLOAK: return new StealthCloakUpdater();
        default: return new NormalArtefactUpdater();
    }
}